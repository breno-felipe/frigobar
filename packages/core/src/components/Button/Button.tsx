import React from 'react';
import styled, { DefaultTheme } from 'styled-components';

interface ButtonProps {
  children: React.ReactNode;
  /** Change background-color */
  skin?: 'primary' | 'info' | 'success' | 'danger' | 'warning' | 'neutral';
  /** change the width, height and font-size values */
  size?: 'small' | 'medium' | 'large' | undefined;
  /** Change disabled state of the button */
  disabled?: boolean;
  /** 100% width */
  full?: boolean;
  /** add border-radius circle */
  rounded?: boolean;
  /** large button */
  large?: boolean;
}

const sizes = ({ spacings }: DefaultTheme) => ({
  small: {
    font: 0.875,
    icon: 1,
    padding: {
      top: spacings.xxsmall,
      right: spacings.small,
      bottom: spacings.xxsmall,
      left: spacings.small,
    },
  },
  medium: {
    font: 1,
    icon: 1.275,
    padding: {
      top: spacings.xsmall,
      right: spacings.medium,
      bottom: spacings.xsmall,
      left: spacings.medium,
    },
  },
  large: {
    font: 1.275,
    icon: 1.5,
    padding: {
      top: spacings.small,
      right: spacings.large,
      bottom: spacings.small,
      left: spacings.large,
    },
  },
});

const Button = styled.button<ButtonProps>`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  transition: all 200ms ease-in-out;
  text-decoration: none;

  ${({
    theme,
    skin = 'neutral',
    size = 'small',
    disabled,
    full,
    rounded,
    large,
  }) => {
    const { radius, colors } = theme;

    return `
      background-color: ${colors[skin][500]};
      color: ${colors.neutral[50]};
      font-size: ${large ? '1.125' : sizes(theme)[size].font}rem;
      padding: ${
        large
          ? `${sizes(theme)[size].padding.top * 2}px
             ${sizes(theme)[size].padding.right * 2}px
             ${sizes(theme)[size].padding.bottom * 2}px
             ${sizes(theme)[size].padding.left * 2}px
          `
          : `${sizes(theme)[size].padding.top}px
             ${sizes(theme)[size].padding.right}px
             ${sizes(theme)[size].padding.bottom}px
             ${sizes(theme)[size].padding.left}px
          `
      };
  
      ${
        disabled
          ? `
        background-color: ${colors.neutral[100]};
        color: ${colors.neutral[400]};
        cursor: not-allowed;
      `
          : `
          &:hover {
            box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
          }
          `
      }
  
      ${full ? 'width: 100%;' : ''}
      ${rounded ? `border-radius: ${radius[3]}px` : ''}
    `;
  }};
`;

Button.defaultProps = {
  skin: 'neutral',
  disabled: false,
  size: 'small',
  full: undefined,
  rounded: undefined,
  large: undefined,
};

export default Button;
