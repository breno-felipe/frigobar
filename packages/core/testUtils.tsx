import React from 'react';
import PropTypes from 'prop-types';
import {
  render,
  fireEvent,
  waitFor,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from './src/theme';

const Providers = ({ children }: { children?: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Providers.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, fireEvent, waitFor };
