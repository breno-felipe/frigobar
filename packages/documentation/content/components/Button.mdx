---
title: Button
menu: components
---

Button component

```jsx react-live=true
<Button>Button</Button>
```
