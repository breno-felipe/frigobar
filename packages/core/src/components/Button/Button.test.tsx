import React from 'react';
import { render } from '../../../testUtils';

import Button from './Button';

describe('<Button />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with all skins', () => {
      const buttons = [
        <Button key="button">Button</Button>,
        <Button size="small" key="small">
          Button
        </Button>,
        <Button size="medium" key="medium">
          Button
        </Button>,
        <Button size="large" key="large">
          Button
        </Button>,
        <Button full key="full">
          Button
        </Button>,
        <Button skin="primary" key="primary">
          Button
        </Button>,
        <Button skin="info" key="info">
          Button
        </Button>,
        <Button skin="success" key="success">
          Button
        </Button>,
        <Button skin="warning" key="warning">
          Button
        </Button>,
        <Button skin="danger" key="danger">
          Button
        </Button>,
        <Button disabled key="disabled">
          Button
        </Button>,
      ];

      buttons.forEach(Component => {
        const { container } = render(Component);
        expect(container).toMatchSnapshot();
      });
    });
  });
});
