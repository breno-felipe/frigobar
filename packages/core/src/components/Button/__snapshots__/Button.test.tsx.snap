// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots should match snapshot with all skins 1`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 2`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 3`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 1rem;
  padding: 8px 16px 8px 16px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 4`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 1.275rem;
  padding: 12px 20px 12px 20px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 5`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
  width: 100%;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 6`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #008d96;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 7`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #25387f;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 8`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #50a47f;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 9`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #cd9447;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 10`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #f44336;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c0:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <button
    class="c0"
  >
    Button
  </button>
</div>
`;

exports[`<Button /> Snapshots should match snapshot with all skins 11`] = `
.c0 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #999;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
  background-color: #f2f2f2;
  color: #a8a8a8;
  cursor: not-allowed;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    Button
  </button>
</div>
`;
